pkg/mimirpb/custom.go

```diff
@@ -38,6 +44,183 @@ func (m *WriteRequest) MinTimestamp() int64 {
+// than maxSize, then the returned partial WriteRequest marshalled size will be bigger than maxSize too.
+//
+// The returned partial WriteRequests are NOT a deep copy of the input one; they may contain references to slices
+// and data from the original WriteRequest.
+// TODO benchmark
+// TODO unit test: consider adding a fuzzy test to stress this logic
+func (m *WriteRequest) SplitByMaxMarshalSize(maxSize int) []*WriteRequest {
+	if m.Size() <= maxSize {
+		return []*WriteRequest{m}
+	}
+
+	partialReqsWithTimeseries := m.splitTimeseriesByMaxMarshalSize(maxSize)
+	partialReqsWithMetadata := m.splitMetadataByMaxMarshalSize(maxSize)
+
+	// Most of the times a write request only have either Timeseries or Metadata.
+	if len(partialReqsWithMetadata) == 0 {
+		return partialReqsWithTimeseries
+	}
+	if len(partialReqsWithTimeseries) == 0 {
+		return partialReqsWithMetadata
+	}
+
+	merged := make([]*WriteRequest, 0, len(partialReqsWithTimeseries)+len(partialReqsWithMetadata))
+	merged = append(merged, partialReqsWithTimeseries...)
+	merged = append(merged, partialReqsWithMetadata...)
+
+	return merged
+}
+
+func (m *WriteRequest) splitTimeseriesByMaxMarshalSize(maxSize int) []*WriteRequest {
+	if len(m.Timeseries) == 0 {
+		return nil
+	}
+
+	newPartialReq := func(preallocTimeseries int) *WriteRequest {
+		return &WriteRequest{
+			Timeseries:              preallocSliceIfNeeded[PreallocTimeseries](preallocTimeseries),
+			Source:                  m.Source,
+			SkipLabelNameValidation: m.SkipLabelNameValidation,
+		}
+	}
+
+	// The partial requests returned by this function will not contain any Metadata,
+	// so we first compute the request size without it.
+	reqSizeWithoutMetadata := m.SizeWithoutMetadata()
+	if reqSizeWithoutMetadata <= maxSize {
+		partialReq := newPartialReq(0)
+		partialReq.Timeseries = m.Timeseries
+		return []*WriteRequest{partialReq}
+	}
+
+	// We assume that different timeseries roughly have the same size (no huge outliers)
+	// so we preallocate the returned slice just adding 1 extra item (+2 because a +1 is to round up).
+	estimatedPartialReqs := (reqSizeWithoutMetadata / maxSize) + 2
+	estimatedTimeseriesPerPartialReq := (len(m.Timeseries) / estimatedPartialReqs) + 2
+
+	partialReq := newPartialReq(estimatedTimeseriesPerPartialReq)
+	partialReqSize := partialReq.Size()
+
+	partialReqs := make([]*WriteRequest, 0, estimatedPartialReqs)
+
+	// Split timeseries into partial write requests.
+	for _, series := range m.Timeseries {
+		seriesSize := series.Size()
+
+		// If the partial request is empty we add the series anyway, in order to avoid an infinite loop
+		// if a single timeseries is bigger than the limit.
+		if partialReqSize+seriesSize > maxSize && !partialReq.IsEmpty() {
+			// The current partial request is full (or close to be full), so we create a new one.
+			partialReqs = append(partialReqs, partialReq)
+			partialReq = newPartialReq(estimatedTimeseriesPerPartialReq)
```

Comment:
Don't we need to reset `partialReqSize` here?

Reply:
Damn yes. It was a bad bug. Fixed in https://github.com/grafana/mimir/pull/8077/commits/0379d8b79d9194a8945bd1d4cd44cd28aae4ab73